pkg/querier/queryrange/roundtrip.go

```diff
@@ -307,11 +308,38 @@ func NewDetectedLabelsTripperware(cfg Config, opts logql.EngineOpts, logger log.
 				base.NewRetryMiddleware(logger, cfg.MaxRetries, metrics.RetryMiddlewareMetrics, namespace),
 			)
 		}
-
-		return NewLimitedRoundTripper(next, l, schema.Configs, queryRangeMiddleware...)
+		limitedRt := NewLimitedRoundTripper(next, l, schema.Configs, queryRangeMiddleware...)
+		return NewDetectedLabelsCardinalityFilter(limitedRt)
 	}), nil
 }
 
+func NewDetectedLabelsCardinalityFilter(rt queryrangebase.Handler) queryrangebase.Handler {
+	return queryrangebase.HandlerFunc(
+		func(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+			res, err := rt.Do(ctx, req)
+			if err != nil {
+				return nil, err
+			}
+
+			resp, ok := res.(*DetectedLabelsResponse)
+			if !ok {
+				return res, nil
+			}
+
+			var result []*logproto.DetectedLabel
+
+			for _, dl := range resp.Response.DetectedLabels {
+				if dl.Cardinality > 2 && dl.Cardinality < 50 {
```

Comment:
I still get confused by our middlewares, so just to make sure, this is happening after we `MergeLabels` right? I wouldn't want to omit results based on partial cardinality of a subquery.

Reply:
yes confirmed again this is executed after MergeLabels. The line `res, err := rt.Do(ctx, req)` is before everything else